docs/architecture/adr-019-protobuf-state-encoding.md

```diff
@@ -110,162 +111,178 @@ type Codec interface {
 
-  return acc.GetAccount(), nil
+```go
+type InterfaceModule interface {
+    RegisterInterfaceTypes(InterfaceRegistry)
 }
 ```
 
-Since the `Codec` implements `auth.Codec` (and all other required interfaces), it is passed to _all_
-the modules and satisfies all the interfaces. Now each module needing to work with interfaces will know
-about all the required types. Note, the use of `interface_type` allows us to avoid a significant
-amount of code boilerplate when implementing the `Codec`.
+The module manager will include a method to call `RegisterInterfaceTypes` on
+every module that implements in order to populate the `InterfaceRegistry`.
 
-A similar concept is to be applied for messages that contain interfaces fields. The module will
-define a "base" concrete message type that the application-level codec will extend via `oneof` that
-fulfills the required message interface.
-
-Example:
+### Using `Any` to encode state
 
-The `MsgSubmitEvidence` defined by the `x/evidence` module contains a field `Evidence` which is an
-interface.
+The SDK will provide support methods `MarshalAny` and `UnmarshalAny` to allow
+easy encoding of state to `Any` in `Codec` implementations. Ex:
 
 ```go
-type MsgSubmitEvidence struct {
-  Evidence  exported.Evidence
-  Submitter sdk.AccAddress
+func (c *Codec) MarshalEvidence(evidenceI eviexported.Evidence) ([]byte, error) {
+	return codec.MarshalAny(evidenceI)
+}
+
+func (c *Codec) UnmarshalEvidence(bz []byte) (eviexported.Evidence, error) {
+	var evi eviexported.Evidence
+	err := codec.UnmarshalAny(c.interfaceContext, &evi, bz)
+	if err != nil {
+		return nil, err
+	}
+	return evi, nil
 }
 ```
 
-Instead, we will implement a "base" message type and an interface which the concrete message type
-must implement.
+### Using `Any` in `sdk.Msg`s
+
+A similar concept is to be applied for messages that contain interfaces fields.
+For example, we can define `MsgSubmitEvidence` as follows where `Evidence` is
+an interface:
 
 ```protobuf
 // x/evidence/types/types.proto
 
-message MsgSubmitEvidenceBase {
+message MsgSubmitEvidence {
   bytes submitter = 1
     [
       (gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"
     ];
+  google.protobuf.Any evidence = 2;
 }
 ```
 
-```go
-// x/evidence/exported/evidence.go
+Note that in order to unpack the evidence from `Any` we do need a reference to
+`InterfaceRegistry`. In order to reference evidence in methods like
+`ValidateBasic` which shouldn't have to know about the `InterfaceRegistry`, we
+introduce an `UnpackInterfaces` phase to deserialization which unpacks
+interfaces before they're needed.
 
-type MsgSubmitEvidence interface {
-  sdk.Msg
+### Unpacking Interfaces
 
-  GetEvidence() Evidence
-  GetSubmitter() sdk.AccAddress
+To implement the `UnpackInterfaces` phase of deserialization which unpacks
+interfaces wrapped in `Any` before they're needed, we create an interface
+that `sdk.Msg`s and other types can implement:
```

Comment:
pp: The `UnpackInterfaces` adds some complexity and may be easily forgotten or done wrong.
I am wondering if it would be easier to have a reflection based approach that recursively scans for fields of types `Any` in the decoded protobuf. There needs to be a cast to the concrete interface or type later anyway. 

Just learned about [`DynamicAny`](https://godoc.org/github.com/gogo/protobuf/types#DynamicAny) type in Gogo. That feels like something that goes into the same direction.

Reply:
This would be possible via reflection if we didn't care about making sure that the packed type was registered for the requested interface. I think there is a subtle difference but my preference is to be more explicit about which types satisfy which interfaces. For instance, say we have two interfaces `Msg` and `PubKey` and some type `MsgX` happens to satisfy both because of either programmer error or pure maliciousness, would we want to allow `MsgX` to be used as a `PubKey`? I would prefer a system that only allowed things that were properly registered.

We could implement all of the `UnpackInterfaces` methods in a way that checks for the right type via a gogo proto plugin. It's not that hard and I've done it before. I was just trying to avoid needing write a new plugin for this. But I can if it's helpful.

`DynamicAny` uses the global registry I'm trying to avoid.

Reply:
> I would prefer a system that only allowed things that were properly registered.
I also prefer this ... portions of the code base that depend on understanding the message format should fail quickly ... other areas that can delegate processing should be able to pass off the message without "opening that box".